Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1288 @@
+                              oldIndex:(NSUInteger)oldIndex
+                              newIndex:(NSUInteger)newIndex
+                              viewItem:(nullable id<ConversationViewItem>)viewItem
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _updateItemType = updateItemType;
+    _oldIndex = oldIndex;
+    _newIndex = newIndex;
+    _viewItem = viewItem;
+
+    return self;
+}
+
+@end
+
+#pragma mark -
+
+@implementation ConversationUpdate
+
+- (instancetype)initWithConversationUpdateType:(ConversationUpdateType)conversationUpdateType
+                                   updateItems:(nullable NSArray<ConversationUpdateItem *> *)updateItems
+                          shouldAnimateUpdates:(BOOL)shouldAnimateUpdates
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _conversationUpdateType = conversationUpdateType;
+    _updateItems = updateItems;
+    _shouldAnimateUpdates = shouldAnimateUpdates;
+
+    return self;
+}
+
++ (ConversationUpdate *)minorUpdate
+{
+    return [[ConversationUpdate alloc] initWithConversationUpdateType:ConversationUpdateType_Minor
+                                                          updateItems:nil
+                                                 shouldAnimateUpdates:NO];
+}
+
++ (ConversationUpdate *)reloadUpdate
+{
+    return [[ConversationUpdate alloc] initWithConversationUpdateType:ConversationUpdateType_Reload
+                                                          updateItems:nil
+                                                 shouldAnimateUpdates:NO];
+}
+
++ (ConversationUpdate *)diffUpdateWithUpdateItems:(nullable NSArray<ConversationUpdateItem *> *)updateItems
+                             shouldAnimateUpdates:(BOOL)shouldAnimateUpdates
+{
+    return [[ConversationUpdate alloc] initWithConversationUpdateType:ConversationUpdateType_Diff
+                                                          updateItems:updateItems
+                                                 shouldAnimateUpdates:shouldAnimateUpdates];
+}
+
+@end
+
+#pragma mark -
+
+// Always load up to n messages when user arrives.
+//
+// The smaller this number is, the faster the conversation can display.
+// To test, shrink you accessability font as much as possible, then count how many 1-line system info messages (our
+// shortest cells) can fit on screen at a time on an iPhoneX
+//
+// PERF: we could do less messages on shorter (older, slower) devices
+// PERF: we could cache the cell height, since some messages will be much taller.
+static const int kYapDatabasePageSize = 25;
+
+// Never show more than n messages in conversation view when user arrives.
+static const int kConversationInitialMaxRangeSize = 300;
+
+// Never show more than n messages in conversation view at a time.
+static const int kYapDatabaseRangeMaxLength = 25000;
+
+static const int kYapDatabaseRangeMinLength = 0;
```

Comment:
Much of the contents of the CVM were drawn from the CVC, without radical changes.  There's a bit more work to use the `delegate` to coordinate certain activities.